### ISSUE TITLE
loss.mean() or loss.sum()

### ISSUE THREAD

**Issue by 704051419** [Alibaba-MIIL/ASL#56]
when i use SGD to optimize the model, should i change loss.sum() to loss.mean()?

**Comment by mrT23** [comment]
this is unrelated to ASL loss

**Comment by thunanguyen** [comment]
> when i use SGD to optimize the model, should i change loss.sum() to loss.mean()?

From my experience that using mean for SGD is far better and the loss is far stable than when using sum. But when using Adam then both sum and mean behave similarly. So depend on the optimizer, you should choose the suitable method

**Comment by ShijianXu** [comment]
Hi, I also have a similar question. Is it possible that using loss.sum() will result in NaN? I train my network on Open Image Dataset, which has more than 10k labels, and I got NaN loss. So I changed it to loss.mean(), and the loss is not NaN, but quite small.

**Comment by mrT23** [comment]
look for bugs of yours. look again. and again.
NaN usually means you have a bug in your code.

you can use mean(), no problem with it, but it won't make the detector learn better during training if you have a bug.
i have been training on open images with sum() all the time

**Comment by thuangb** [comment]
> Hi, I also have a similar question. Is it possible that using loss.sum() will result in NaN? I train my network on Open Image Dataset, which has more than 10k labels, and I got NaN loss. So I changed it to loss.mean(), and the loss is not NaN, but quite small.

This could be many things, usually bugs in computing gradient. You can use torch.autograd.set_detect_anomaly(True) to detect where the gradient goes wrong

**Comment by SippieCup** [comment]
a NaN can occur after awhile of training under certain conditions, such as when a classification does not have enough examples in the training data and is using loss.mean() (it may also require mixed precision).

It occurs when [you calculate probabilities from the logits](https://github.com/Alibaba-MIIL/ASL/blob/7114637713619f01906ed73bbfa182565cd3a77d/src/loss_functions/losses.py#L24) and the sigmoid goes to -inf.

You can "fix" it by clamping the gradients/logits/whatever themselves with a minimum value, but that would end up with that particular label never being able to be learned.


The best solution is to fix your dataset by adding more instances of the problem classes if possible, and using loss.sum() instead of mean.

**Comment by SmileTAT** [comment]
> a NaN can occur after awhile of training under certain conditions, such as when a classification does not have enough examples in the training data and is using loss.mean() (it may also require mixed precision).
> 
> It occurs when [you calculate probabilities from the logits](https://github.com/Alibaba-MIIL/ASL/blob/7114637713619f01906ed73bbfa182565cd3a77d/src/loss_functions/losses.py#L24) and the sigmoid goes to -inf.
> 
> You can "fix" it by clamping the gradients/logits/whatever themselves with a minimum value, but that would end up with that particular label never being able to be learned.
> 
> The best solution is to fix your dataset by adding more instances of the problem classes if possible, and using loss.sum() instead of mean.

Could you write an example whose loss goes to -inf? Thanks.
